From a8423fb1a178b53ccccb7bd1fccd03c4942839c6 Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Tue, 26 Aug 2014 14:31:05 +0200
Subject: [PATCH 052/118] daemon: add add_sprintf

Bring the add_sprintf function for stringsbuf from the library.

(cherry picked from commit 9b501fa5c6365958de362ecaf0f4e90c147978a2)
---
 daemon/daemon.h   |  2 ++
 daemon/guestfsd.c | 20 ++++++++++++++++++++
 2 files changed, 22 insertions(+)

diff --git a/daemon/daemon.h b/daemon/daemon.h
index 544feb2..f2e3c7a 100644
--- a/daemon/daemon.h
+++ b/daemon/daemon.h
@@ -70,6 +70,8 @@ struct stringsbuf {
  */
 extern int add_string_nodup (struct stringsbuf *sb, char *str);
 extern int add_string (struct stringsbuf *sb, const char *str);
+extern int add_sprintf (struct stringsbuf *sb, const char *fs, ...)
+  __attribute__((format (printf,2,3)));
 extern int end_stringsbuf (struct stringsbuf *sb);
 
 extern size_t count_strings (char *const *argv);
diff --git a/daemon/guestfsd.c b/daemon/guestfsd.c
index bba13f8..e78dd4d 100644
--- a/daemon/guestfsd.c
+++ b/daemon/guestfsd.c
@@ -476,6 +476,26 @@ add_string (struct stringsbuf *sb, const char *str)
 }
 
 int
+add_sprintf (struct stringsbuf *sb, const char *fs, ...)
+{
+  va_list args;
+  char *str;
+  int r;
+
+  va_start (args, fs);
+  r = vasprintf (&str, fs, args);
+  va_end (args);
+  if (r == -1) {
+    reply_with_perror ("vasprintf");
+    free_stringslen (sb->argv, sb->size);
+    sb->argv = NULL;
+    return -1;
+  }
+
+  return add_string_nodup (sb, str);
+}
+
+int
 end_stringsbuf (struct stringsbuf *sb)
 {
   return add_string_nodup (sb, NULL);
-- 
2.7.4

